src/dataflow-types/src/types.rs

```diff
@@ -588,14 +593,36 @@ pub struct ProtobufEncoding {
     pub message_name: String,
 }
 
-/// Encoding in CSV format, with `n_cols` columns per row, with an optional header.
+/// Encoding in CSV format
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub struct CsvEncoding {
-    pub header_row: bool,
-    pub n_cols: usize,
+    pub columns: ColumnSpec,
     pub delimiter: u8,
 }
 
+impl CsvEncoding {
+    pub fn has_header_rows(&self) -> bool {
+        matches!(self.columns, ColumnSpec::HeaderNames(_))
+    }
+}
+
+/// What we know about the CSV columns
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum ColumnSpec {
+    Count(usize),
+    HeaderNames(Vec<String>),
+}
+
+impl ColumnSpec {
+    /// Either the length of the names or the size of the count
```

Comment:
I think this wants to read more like "The number of columns described by the column spec." As it stands it is describing what the method does rather than what the return value means.

Reply:
fair